R/textstat_readability.R

```diff
@@ -526,9 +526,11 @@ textstat_readability.corpus <- function(x,
 
     x <- texts(x)
     if (!is.null(min_sentence_length) || !is.null(max_sentence_length)) {
-        x <- char_trim(x, "sentences",
-                       min_ntoken = min_sentence_length,
-                       max_ntoken = max_sentence_length)
+        temp <- char_trim(x, "sentences",
+                          min_ntoken = min_sentence_length,
+                          max_ntoken = max_sentence_length)
+        x[names(temp)] <- temp
```

Comment:
This could be done in a more robust way in corpus using factor levels of docid etc.

Reply:
Yes it could, and the root of the problem is deeper, but here it's just a fix to prevent documents from disappearing. The root problem comes when we reshape to sentences, which obliterates empty documents, and then they cannot be reshaped to documents. I thought we (you 😀) could fix this later.

``` r
library("quanteda")
## Package version: 2.1.0
## Parallel computing: 2 of 8 threads used.
## See https://quanteda.io for tutorials and examples.
## 
## Attaching package: 'quanteda'
## The following object is masked from 'package:utils':
## 
##     View

corp <- corpus(c(
  d1 = "The cat in the hat at green ham and eggs.",
  d2 = "",
  d3 = "Once upon a time."
))
corp
## Corpus consisting of 3 documents.
## d1 :
## "The cat in the hat at green ham and eggs."
## 
## d2 :
## ""
## 
## d3 :
## "Once upon a time."

corp %>%
  corpus_reshape(to = "sentences") %>%
  corpus_reshape(to = "documents")
## Corpus consisting of 2 documents.
## d1 :
## "The cat in the hat at green ham and eggs."
## 
## d3 :
## "Once upon a time."
```

Reply:
That was my guess. If you postpone solving problems, they just become bigger.

You could use `segment_texts()` which is the internal for `corpus_reshape()`:
```r
segment_texts <- function(x, pattern = NULL, valuetype = "regex",
                          case_insensitive = TRUE,
                          extract_pattern = FALSE, pattern_position = "after",
                          omit_empty = TRUE, what = "other", ...) {
```
Interestingly, `omit_empty = FALSE` here...

https://github.com/quanteda/quanteda/blob/c43c37b29e581674f0bd0ed503d02b75e84fda7f/R/corpus_reshape.R#L68-L69

Reply:
Actually the problem is in https://github.com/quanteda/quanteda/pull/1977/files#diff-745460f071607a11357036687422228dR50, where the `corpus_subset()` removes documents - here, sentences - that don't meet the conditions. Adding the code in `textstat_readability()` as I have ensures that no document is ever dropped from the result. We can open a separate issue for what to do with operations that result in empty (`""`) or `NA` documents, but I think ensuring that all documents are returned here - meaning https://github.com/quanteda/quanteda/pull/1977/files#diff-745460f071607a11357036687422228dR50 -  in is still appropriate.

Reply:
The links do not seem to be working. Can you link to the code in the master?

Reply:
OK will do. The first link was 
https://github.com/quanteda/quanteda/blob/c43c37b29e581674f0bd0ed503d02b75e84fda7f/R/corpus_trim.R#L50
where the subset operation removes the "document". So a one-sentence document that is shorter than the minimum readability tokens is simply removed. The extra code in `textstat_readability()` ensures this will never happen.

But I will also open a larger issue about what happens to documents during reshape operations etc. when those documents are empty.

The second link was just referring to the same code at the top of this comment, i.e.
```r
    x[names(temp)] <- temp
```